core/lib/contract_verifier/src/resolver/github/mod.rs

```diff
@@ -0,0 +1,279 @@
+        self.zkvyper_versions = gh_client
+            .zkvyper_versions()
+            .await
+            .context("failed fetching zkvyper versions")?;
+        self.last_updated = Instant::now();
+        Ok(())
+    }
+
+    async fn update_if_needed(&mut self, gh_client: &GitHubApi) -> anyhow::Result<()> {
+        if self.is_outdated() {
+            tracing::info!("GH compiler versions cache outdated, updating");
+            self.update(gh_client).await?;
+        }
+        Ok(())
+    }
+}
+
+impl GitHubCompilerResolver {
+    pub async fn new() -> anyhow::Result<Self> {
+        let artifacts_dir = tempfile::tempdir().context("failed creating temp dir")?;
+        let gh_client = GitHubApi::new();
+        let mut supported_versions = SupportedVersions::default();
+        if let Err(err) = supported_versions.update(&gh_client).await {
+            // We don't want the resolver to fail at creation if versions can't be fetched.
+            // It shouldn't bring down the whole application, so the expectation here is that
+            // the versions will be fetched later.
+            tracing::error!("failed syncing compiler versions at start: {:?}", err);
+        }
+
+        Ok(Self {
+            artifacts_dir,
+            gh_client,
+            client: reqwest::Client::new(),
+            supported_versions: RwLock::new(supported_versions),
+        })
+    }
+}
+
+impl GitHubCompilerResolver {
+    async fn download_version_if_needed(
+        &self,
+        compiler: CompilerType,
+        version: &str,
+    ) -> anyhow::Result<()> {
+        if compiler.exists(self.artifacts_dir.path(), version).await? {
+            tracing::debug!("Compiler {}:{} exists", compiler.as_str(), version);
+            return Ok(());
+        }
+
+        tracing::info!("Downloading {}:{}", compiler.as_str(), version);
+        let lock = self.supported_versions.read().await;
+        let versions = match compiler {
+            CompilerType::Solc => &lock.solc_versions,
+            CompilerType::ZkSolc => &lock.zksolc_versions,
+            CompilerType::Vyper => &lock.vyper_versions,
+            CompilerType::ZkVyper => &lock.zkvyper_versions,
+        };
+
+        let version_url = versions
+            .get(version)
+            .ok_or_else(|| {
+                ContractVerifierError::UnknownCompilerVersion("solc", version.to_owned())
+            })?
+            .clone();
+        drop(lock);
+        let path = compiler.bin_path_unchecked(self.artifacts_dir.path(), version);
+
+        let response = self.client.get(version_url).send().await?;
+        let body = response.bytes().await?;
+
+        tracing::info!("Saving {}:{} to {:?}", compiler.as_str(), version, path);
+
+        tokio::fs::create_dir_all(path.parent().unwrap())
+            .await
+            .map_err(|e| anyhow::anyhow!("failed to create dir: {}", e))?;
+
+        let mut file = tokio::fs::File::create(path)
```

Comment:
You're right, I forgot to add the protection for that. It's not a concern right now, because AFAIU contract verifier is single-threaded, but it doesn't justify potentially incorrect behavior. 

I'll also replaced `File::create` with `File::create_new` to make sure that two streams won't write to the same file.